composer/utils/eval_client/local_eval_client.py

```diff
@@ -42,27 +51,87 @@ def update_offline_helper(self, code_gen: str, test_input: str, test_output: str
         """
-        mod = types.ModuleType('test_module')
-        val.value = 0
-        result = None
-        expected_result = None
-        try:
-            exec(code_gen, mod.__dict__)
-
-            result = mod.__dict__[entry_point](*eval(test_input))
-            syntax_compiled = True
+        if language == 'python':
+            mod = types.ModuleType('test_module')
+            val.value = 0
+            result = None
+            expected_result = None
             try:
-                expected_result = eval(test_output)
-            except:
-                expected_result = test_output
+                exec(code_gen, mod.__dict__)
+
+                result = mod.__dict__[entry_point](*eval(test_input))
+                syntax_compiled = True
+                try:
+                    expected_result = eval(test_output)
+                except:
+                    expected_result = test_output
+
+            except Exception as _:
+                syntax_compiled = False  # don't print out warning b/c this can happen frequently
+
+            if syntax_compiled:
+                val.value = int(result == expected_result)
+            else:
+                val.value = 0
+        elif language == 'c++':
+            prefix = '#include <iostream>\n#include <vector>\n#include <string>\n#include <map>\n#include <math.h>\nusing namespace std;\nbool issame(vector<string> a,vector<string>b){\n    if (a.size()!=b.size()) return false;\n    for (int i=0;i<a.size();i++)\n    {\n    if (a[i]!=b[i]) return false;\n    }\n    return true;\n}\nbool issame(float a, float b){\n    return abs(a - b) < 1e-4;\n}\nbool issame(bool a, bool b){\n    return a == b;\n}\nbool issame(int a, int b){\n    return a == b;\n}\nbool issame(vector<int> a,vector<int>b){\n    if (a.size()!=b.size()) return false;\n    for (int i=0;i<a.size();i++)\n    {\n        if (a[i]!=b[i]) return false;\n    }\n    return true;\n}\nbool issame(string a, string b){\n    return a == b;\n}\nbool issame(vector<float> a,vector<float>b){\n    if (a.size()!=b.size()) return false;\n    for (int i=0;i<a.size();i++)\n    {\n        if (abs(a[i]-b[i])>1e-4) return false;\n    }\n    return true;\n}\nbool issame(double a, double b){\n    return abs(a - b) < 1e-3;\n}\nbool issame(vector<vector<int>> a,vector<vector<int>> b){\n    if (a.size()!=b.size()) return false;\n\n    for (int i=0;i<a.size();i++)\n    {\n        if (a[i].size()!=b[i].size()) return false;\n        for (int j=0;j<a[i].size();j++)\n            if (a[i][j]!=b[i][j]) return false;\n    }\n    return true;\n}\nbool issame(map<char,int> a,map<char,int> b){\n    if (a.size()!=b.size()) return false;\n    map <char,int>::iterator it;\n    for (it=a.begin();it!=a.end();it++)\n    {\n        char w1=it->first;\n        int w2=it->second;\n        if (b.find(w1)==b.end()) return false;\n        if (b[w1]!=w2) return false;\n    }\n\n    return true;\n}\nbool issame(long long a, long long b){\n    return a == b;\n}\n'
```

Comment:
Please include a comment detailing what this is

Is there a way to make it more readable...? like textwrap deedent or something?

Reply:
done